physics/hierarchical_system_body.hpp

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include "physics/hierarchical_system.hpp"
+
+#include <iterator>
+
+namespace principia {
+namespace physics {
+
+template<typename Frame>
+HierarchicalSystem<Frame>::HierarchicalSystem(
+    not_null<std::unique_ptr<MassiveBody const>> primary)
+    : system_(std::move(primary)) {
+  systems_[system_.primary.get()] = &system_;
+}
+
+template<typename Frame>
+void HierarchicalSystem<Frame>::Add(
+    not_null<std::unique_ptr<MassiveBody const>> body,
+    not_null<MassiveBody const*> const parent,
+    KeplerianElements<Frame> const& jacobi_osculating_elements) {
+  not_null<MassiveBody const*> unowned_body = body.get();
+  System& parent_system = *systems_[parent];
+  parent_system.satellites.emplace_back(
+      make_not_null_unique<Subsystem>(std::move(body)));
+  {  // Hide the moved-from |body|.
+    not_null<MassiveBody const*> body = unowned_body;
+    not_null<Subsystem*> inserted_system =
+        parent_system.satellites.back().get();
+    systems_[body] = inserted_system;
+    inserted_system->jacobi_osculating_elements = jacobi_osculating_elements;
+  }
+}
+
+template<typename Frame>
+typename HierarchicalSystem<Frame>::BarycentricSystem
+HierarchicalSystem<Frame>::Get() {
+  auto const semimajor_axis_less_than = [](
+      not_null<std::unique_ptr<Subsystem>> const& left,
+      not_null<std::unique_ptr<Subsystem>> const& right) -> bool {
+    return left->jacobi_osculating_elements.semimajor_axis <
+           right->jacobi_osculating_elements.semimajor_axis;
+  };
+
+  // Data about a |Subsystem|.
+  struct BarycentricSubystem {
+    // A |MassiveBody| with the mass of the whole subsystem.
+    std::unique_ptr<MassiveBody> equivalent_body;
+    // The bodies composing the subsystem.
+    std::vector<not_null<std::unique_ptr<MassiveBody const>>> bodies;
+    // Their |DegreesOfFreedom| with respect to the barycentre of the subsystem.
+    std::vector<RelativeDegreesOfFreedom<Frame>> barycentric_degrees_of_freedom;
+  };
+
+  std::function<BarycentricSubystem(System&)> to_barycentric =
```

Comment:
I am not in love with this gigantic lambda.  Moving it to a static function would make it easier to understand what it accesses and what its role is (it could, for instance, have comments, imagine that).

I am actually wondering if this should be two functions, one computing |satellite_degrees_of_freedom|, |jacobi_coordinates|, and |result.bodies| and the other filling |result.barycentric_degrees_of_freedom|.

Anyway this is so complicated that I only have the faintest idea what it does.